app/src/main/java/com/github/mateo762/myapplication/upload_gallery/UploadPictureActivity.kt

```diff
@@ -0,0 +1,160 @@
+
+    private lateinit var binding: UploadPictureBinding
+    private lateinit var imageUri: Uri
+
+    private var defaultUserString: String = "testUser"
+    private var currentUser: String = defaultUserString
+
+    lateinit var imagesList: ArrayList<UserImage>
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        binding = UploadPictureBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+
+        binding.selectPictureButton.setOnClickListener {
+            selectImage()
+        }
+
+        binding.uploadPictureButton.setOnClickListener {
+            uploadImage()
+        }
+
+        binding.imageRecycler.layoutManager = LinearLayoutManager(this)
+        imagesList = arrayListOf()
+
+        //Populate the recycler view with the downloaded images from the current user firebase database
+        var currentUser = Firebase.auth.currentUser?.displayName
+
+        if (currentUser == null) {
+            currentUser = defaultUserString
+        }
+
+        val db = Firebase.database.reference.child("users").child(currentUser)
+        db.addValueEventListener(object : ValueEventListener {
+            override fun onDataChange(snapshot: DataSnapshot) {
+                imagesList = arrayListOf()
+                if (snapshot.exists()) {
+                    for (dataSnapshot in snapshot.children) {
+                        val image = dataSnapshot.getValue(UserImage::class.java)
+                        imagesList.add(image!!)
+                    }
+
+                    binding.imageRecycler.adapter =
+                        ImageAdapter(imagesList, this@UploadPictureActivity)
+                }
+            }
+
+            override fun onCancelled(error: DatabaseError) {
+                Toast.makeText(this@UploadPictureActivity, error.toString(), Toast.LENGTH_SHORT)
+                    .show()
+            }
+        })
+
+
+    }
+
+    //Select an image from the gallery
+    private fun selectImage() {
+        val intent = Intent()
+        intent.type = "image/*"
+        intent.action = Intent.ACTION_GET_CONTENT
+
+        startActivityForResult(intent, 100)
+    }
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        super.onActivityResult(requestCode, resultCode, data)
+
+        if (requestCode == 100 && resultCode == RESULT_OK) {
+            imageUri = data?.data!!
+            binding.selectedImagePreview.setImageURI(imageUri)
+        }
+    }
+
+    //Upload an image to the Firebase Storage and then save under the current user database
+    private fun uploadImage() {
+
+        val progressDialog = ProgressDialog(this)
+
+        progressDialog.setMessage("Uploading File..")
+        progressDialog.setCancelable(false)
+        progressDialog.show()
+
+        //Generate a file name based on the upload time
+        val formatter = SimpleDateFormat("yyy_MM_dd_HH_mm_ss", Locale.getDefault())
```

Comment:
I would prefer if the hardcoded values are in the const val parameter, in order not to have magic number and strings, but it is not a big deal in this case